api/valor_api/backend/query/solvers.py

```diff
@@ -1,442 +1,690 @@
-    query = query_solver(
-        args=args,
-        selected=selected,
-        filtered=qset,
-        expressions=expressions,
-    )
-    sub_qset = filtered.intersection(unique_set).union({Datum})
-    subquery = (
-        subquery_solver(
-            args=[Datum.id],
-            selected={Datum},
-            filtered=sub_qset,
-            expressions=expressions,
+    Returns a list of join operations that connect the 'select_from' table to all the provided 'tables'.
+
+    Parameters
+    ----------
+    select_from : TableTypeAlias
+        The table that is being selected from.
+    label_source : LabelSourceAlias
+        The table that is being used as the source of labels.
+    tables : set[TableTypeAlias]
+        The set of tables that need to be joined.
+
+    Returns
+    -------
+    list[Callable]
+        An ordered list of join operations.
+    """
+    if label_source not in {GroundTruth, Prediction, Annotation}:
+        raise ValueError(
+            "Label source must be either GroundTruth, Prediction or Annotation."
         )
-        if pivot_table in filtered or sub_qset != {pivot_table}
-        else None
-    )
-    return query, subquery
-
-
-def solve_graph(
-    select_args: tuple[
-        TableTypeAlias | InstrumentedAttribute | UnaryExpression
-    ],
-    selected_tables: set[TableTypeAlias],
-    filter_by_tables: set[TableTypeAlias],
-    expressions: dict[
-        TableTypeAlias, list[ColumnElement[bool] | BinaryExpression]
-    ],
-    pivot_table: TableTypeAlias | None = None,
-) -> tuple[Select | None, None]:
+
+    table_mapping = map_label_source_to_neighbor_tables[label_source]
+    join_mapping = map_label_source_to_neighbor_joins[label_source]
+
+    ordered_tables = [select_from]
+    ordered_joins = []
+    for target in tables:
+        if select_from is target:
+            continue
+        solution = _recursive_search(
+            table=select_from,
+            target=target,
+            mapping=table_mapping,
+        )
+        for idx in range(1, len(solution)):
+            lhs = solution[idx - 1]
+            rhs = solution[idx]
+            if rhs not in ordered_tables:
+                ordered_tables.append(rhs)
+                ordered_joins.append(join_mapping[lhs][rhs])
+    return ordered_joins
+
+
+def generate_query(
+    select_statement: Select[Any] | Query[Any],
+    args: tuple[TableTypeAlias | InstrumentedAttribute | UnaryExpression],
+    select_from: TableTypeAlias,
+    label_source: LabelSourceAlias,
+    filter_: Filter | None = None,
+) -> Select[Any] | Query[Any]:
     """
-    Selects best fitting graph to run query generation and returns tuple(query, subquery | None).
+    Generates the main query.
 
-    Description
+    Includes all args-related and filter-related tables.
+
+    Parameters
     ----------
-    To construct complex queries it is necessary to describe the relationship between predictions and groundtruths.
```

Comment:
nit: could we update and add back these longer docstrings somewhere? I found them helpful